src/tiled/changemapobject.cpp

```diff
@@ -203,3 +203,63 @@ void DetachObjects::undo()
 
     emit mMapDocument->mapObjectModel()->objectsChanged(mMapObjects);
 }
+
+ResetInstances::ResetInstances(MapDocument *mapDocument,
+                               QList<MapObject *> mapObjects,
+                               QUndoCommand *parent)
+    : QUndoCommand(QCoreApplication::translate("Undo Commands",
+                                               "Reset %n Instnaces",
+                                               nullptr, mapObjects.size()), parent)
+    , mMapDocument(mapDocument)
+    , mMapObjects(mapObjects)
+{
+    for (const MapObject *object : mMapObjects) {
+        mProperties.append(object->properties());
+        mChangedProperties.append(object->changedProperties());
+        mOldMapObjects.append(object->clone());
+    }
+}
+
+ResetInstances::~ResetInstances()
+{
+    qDeleteAll(mOldMapObjects);
+}
+
+void ResetInstances::redo()
+{
+    for (auto object : mMapObjects) {
+        // Template instances initially don't hold any custom properties
+        object->clearProperties();
+
+        // Reset built-in properties
+        object->setChangedProperties(0);
+        object->syncWithTemplate();
+    }
+
+    emit mMapDocument->objectsChanged(mMapObjects);
+
+    // This signal forces updating custom properties in the properties dock
+    emit mMapDocument->selectedObjectsChanged();
+}
+
+void ResetInstances::undo()
+{
+    // TODO: Is there a better way than saving clones and resetting like this?
```

Comment:
Maybe, but I think this is fine.

Reply:
Alright, I found out that I didn't update name or type, I will add them.

Reply:
Also updating the `templateRef` was unnecessary.